lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,535 @@
+use crate::partition::Partitioner;
+use crate::time::Timer;
+use crate::ByteSizeOf;
+use futures::stream::Stream;
+use pin_project::pin_project;
+use std::collections::HashMap;
+use std::hash::BuildHasherDefault;
+use std::mem;
+use std::num::NonZeroUsize;
+use std::pin::Pin;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use tokio::time::{interval, Interval, MissedTickBehavior};
+use twox_hash::XxHash64;
+
+/// A [`Timer`] for [`Batch`]
+///
+/// A `Batch` implementation must be rigged to flush every so often on a
+/// timer. Absent any particular concerns this timer will serve most
+/// purposes. It stores an internal notion of when it was started and how "wide"
+/// away from that start time can drift before the timer has elapsed.
+pub struct BatcherTimer {
+    interval: Interval,
+}
+
+impl BatcherTimer {
+    #[allow(dead_code)]
+    pub fn new(period: Duration) -> Self {
+        let mut interval = interval(period);
+        interval.set_missed_tick_behavior(MissedTickBehavior::Delay);
```

Comment:
My gut feeling is that we _may_ want the "burst" mode: if we have a slowdown, we should try to flush out older data as soon as possible until we're back on track.  While "delay" mode is more consistent, slowdowns end up messing with our "data should never go out _later_ than the batch timeout" invariant: the batches themselves will be consistent, but data itself that is stuck behind slow data itself will violate that invariant.

Thoughts?

Reply:
You know, it's not clear to me. I think what I'll do is make this configurable by the caller and then we can measure in practice.

Reply:
Eh, actually, thinking on this more I believe `Delay` is the right call. The timer here is not actually responsible for flushing historical data so much as it is a backstop to keep slow partitions from never flushing. If an elapsed check gets missed it's not going to cause a backup, I don't think. Something worth measuring and/or modeling but I'm content with this for now.